nats-compiler-plugin/src/main/java/io/ballerina/stdlib/nats/plugin/NatsFunctionValidator.java

```diff
@@ -0,0 +1,223 @@
+                                            FunctionDefinitionNode functionDefinitionNode) {
+        if (parameters.size() > 0) {
+            validateFirstParam(functionDefinitionNode, parameters.get(0));
+        }
+        if (parameters.size() < 1) {
+            context.reportDiagnostic(PluginUtils.getDiagnostic(CompilationErrors.MUST_HAVE_MESSAGE,
+                    DiagnosticSeverity.ERROR, functionDefinitionNode.functionSignature().location()));
+        }
+        if (parameters.size() > 1) {
+            context.reportDiagnostic(PluginUtils.getDiagnostic(CompilationErrors.ONLY_PARAMS_ALLOWED,
+                    DiagnosticSeverity.ERROR, functionDefinitionNode.functionSignature().location()));
+        }
+    }
+
+    private void validateOnErrorFunctionParameters(SeparatedNodeList<ParameterNode> parameters,
+                                                   FunctionDefinitionNode functionDefinitionNode) {
+        if (parameters.size() > 1) {
+            validateFirstParam(functionDefinitionNode, parameters.get(0));
+            validateErrorParam(functionDefinitionNode, parameters.get(1));
+        }
+        if (parameters.size() < 2) {
+            context.reportDiagnostic(PluginUtils.getDiagnostic(CompilationErrors.MUST_HAVE_MESSAGE_AND_ERROR,
+                    DiagnosticSeverity.ERROR, functionDefinitionNode.functionSignature().location()));
+        }
+        if (parameters.size() > 2) {
+            context.reportDiagnostic(PluginUtils.getDiagnostic(CompilationErrors.ONLY_PARAMS_ALLOWED_ON_ERROR,
+                    DiagnosticSeverity.ERROR, functionDefinitionNode.functionSignature().location()));
+        }
+    }
+
+    private void validateFirstParam(FunctionDefinitionNode functionDefinitionNode,
+                                    ParameterNode parameterNode) {
+        RequiredParameterNode requiredParameterNode = (RequiredParameterNode) parameterNode;
+        Node parameterTypeNode = requiredParameterNode.typeName();
+        if (!parameterTypeNode.kind().equals(SyntaxKind.QUALIFIED_NAME_REFERENCE)) {
+            context.reportDiagnostic(PluginUtils.getDiagnostic(
+                    CompilationErrors.INVALID_FUNCTION_PARAM_MESSAGE,
+                    DiagnosticSeverity.ERROR, functionDefinitionNode.location()));
+        } else {
+            Token modulePrefix = ((QualifiedNameReferenceNode) parameterTypeNode).modulePrefix();
+            IdentifierToken identifierToken = ((QualifiedNameReferenceNode) parameterTypeNode).identifier();
+            if (!modulePrefix.text().equals(PluginConstants.PACKAGE_PREFIX) ||
+                    !identifierToken.text().equals(PluginConstants.MESSAGE)) {
+                context.reportDiagnostic(PluginUtils.getDiagnostic(
+                        CompilationErrors.INVALID_FUNCTION_PARAM_MESSAGE,
+                        DiagnosticSeverity.ERROR, functionDefinitionNode.location()));
+            }
```

Comment:
A couple of questions/issues regarding this
- I think it's best if the semantic API is used for these sorta validations. For example, what if the user uses an alias for the module prefix? e.g., `import ballerina/nats as foo;`
- Also what if the expected param type is a subtype of the param type user has specified? i.e., the method is a subtype of the expected method type.

Reply:
Can you give an example for the "expected param type is a subtype of the param type user has specified"

Reply:
I meant, let's say the function signature is expected to look like the following:
```ballerina
function (T1|T2 param);
```
But the user defines the following function instead:
```ballerina
function foo(T1 p) {}
```
Is this a valid scenario in the above case?

Reply:
NATS doesn't have union types allowed in any of the the remote methods in services. Only valid signature is 
```ballerina
remote function onMessage(nats:Message msg) {}
```